ccf_specs.md

```diff
@@ -576,72 +609,104 @@ composite-type = [
     ]
 ]
 
+composite-interface-type = [
+    composite-type-id
+]
+
 struct-type =
     ; cbor-tag-struct-type
-    #6.137(composite-type)
+    #6.160(composite-type)
 
 resource-type =
     ; cbor-tag-resource-type
-    #6.138(composite-type)
-
-contract-type =
-    ; cbor-tag-contract-type
-    #6.140(composite-type)
+    #6.161(composite-type)
 
 event-type =
     ; cbor-tag-event-type
-    #6.139(composite-type)
+    #6.162(composite-type)
+
+contract-type =
+    ; cbor-tag-contract-type
+    #6.163(composite-type)
 
 enum-type =
     ; cbor-tag-enum-type
-    #6.141(composite-type)
+    #6.164(composite-type)
+
+struct-interface-type =
+    ; cbor-tag-struct-interface-type
+    #6.176(composite-interface-type)
+
+resource-interface-type =
+    ; cbor-tag-resource-interface-type
+    #6.177(composite-interface-type)
+
+contract-interface-type =
+    ; cbor-tag-contract-interface-type
+    #6.178(composite-interface-type)
 
 inline-type =
     simple-type
     / optional-type
     / varsized-array-type
     / constsized-array-type
     / dict-type
+    / reference-type
+    / restricted-type
     / capability-type
     / type-ref
 
 simple-type =
     ; cbor-tag-simple-type
-    #6.132(simple-type-id)
+    #6.137(simple-type-id)
 
 optional-type =
     ; cbor-tag-optional-type
-    #6.133(inline-type)
+    #6.138(inline-type)
 
 varsized-array-type =
     ; cbor-tag-varsized-array-type
-    #6.134(inline-type)
+    #6.139(inline-type)
 
 constsized-array-type =
     ; cbor-tag-constsized-array-type
-    #6.135([
+    #6.140([
         array-size: uint,
         element-type: inline-type
     ])
 
 dict-type =
     ; cbor-tag-dict-type
-    #6.136([
+    #6.141([
         key-type: inline-type,
         element-type: inline-type
     ])
 
+reference-type =
+    ; cbor-tag-reference-type
+    #6.142([
+      authorized: bool,
+      type: inline-type,
+    ])
+
+restricted-type =
+    ; cbor-tag-restricted-type
+    #6.143([
+      type: inline-type,
+      restrictions: [+ inline-type]
+    ])
+
 capability-type =
     ; cbor-tag-capability-type
     ; use an array as an extension point
-    #6.142([
+    #6.144([
         ; borrow-type
         inline-type
     ])
 
 type-ref =
     ; cbor-tag-type-ref
-    #6.131(
+    #6.136(
         ; composite-type-id
```

Comment:
This comment is now a bit confusing, as `composite-type-id` is now a rule. 

`type-ref` is a reference to type information in this encoding, not a Cadence type ID (see above: location + qualified identifier)

Reply:
Good point!  

I made the following changes in commit 13a5309d6e6cf683068470c68b0e9add21493185:
- defined `id` as a unique identifier used by CCF to associate composite/interface type information with value through type-ref and type-value-ref.
- defined `cadence-type-id` as an identifier used by Cadence to identify types.
- used `id` in composite/interface type definitions, `type-ref`, and `type-value-ref`
- used `cadence-type-id` in other places for clarification